Add ElementIndex

wthrowe opened the pull request.

wthrowe committed:
Add ElementIndex

nilsdeppe commented on src/Domain/ElementIndex.hpp:
```diff
@@ -0,0 +1,76 @@
+};
+}  // namespace ElementIndex_detail
+
+/// A class for indexing a Charm array by Element.
```
We should probably add this to the `Parallel` doxygen group

nilsdeppe commented on src/Domain/ElementIndex.cpp:
```diff
@@ -0,0 +1,107 @@
+  : block_id_(block_id), refinement_level_(segment_id.refinement_level()),
+    index_(segment_id.index()) {
+  ASSERT(block_id < two_to_the(block_id_bits),
+         "Block id out of bouunds: " << block_id);
```
bouunds -> bounds

kidder reviewed (commented):
fix and squash

wthrowe commented:
In addition to the requested changes I adjusted the refinement level assertion to check that it was <= max_refinement_level instead of just that it fit in the allocated bits and modified the associated message to print the out-of-range value before it was truncated to the bitfield.
